google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/internal/AssigningSubscriber.java

```diff
@@ -0,0 +1,123 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsublite.cloudpubsub.internal;
+
+import static com.google.cloud.pubsublite.internal.ExtractStatus.toCanonical;
+import static com.google.cloud.pubsublite.internal.Preconditions.checkState;
+
+import com.google.cloud.pubsublite.Partition;
+import com.google.cloud.pubsublite.cloudpubsub.Subscriber;
+import com.google.cloud.pubsublite.internal.CloseableMonitor;
+import com.google.cloud.pubsublite.internal.ProxyService;
+import com.google.cloud.pubsublite.internal.wire.Assigner;
+import com.google.cloud.pubsublite.internal.wire.AssignerFactory;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import io.grpc.StatusException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class AssigningSubscriber extends ProxyService implements Subscriber {
+  private final PartitionSubscriberFactory subscriberFactory;
+
+  private final CloseableMonitor monitor = new CloseableMonitor();
+
+  @GuardedBy("monitor.monitor")
+  private final Map<Partition, Subscriber> liveSubscriberMap = new HashMap<>();
+
+  @GuardedBy("monitor.monitor")
+  private final List<Subscriber> stoppingSubscribers = new ArrayList<>();
+
+  @GuardedBy("monitor.monitor")
+  private boolean shutdown = false;
+
+  public AssigningSubscriber(
+      PartitionSubscriberFactory subscriberFactory, AssignerFactory assignerFactory)
+      throws StatusException {
+    this.subscriberFactory = subscriberFactory;
+    Assigner assigner = assignerFactory.New(this::handleAssignment);
+    addServices(assigner);
+  }
+
+  @Override
+  protected void start() {}
+
+  @Override
+  protected void stop() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      shutdown = true;
+      liveSubscriberMap.values().forEach(subscriber -> subscriber.stopAsync().awaitTerminated());
+      liveSubscriberMap.clear();
+      stoppingSubscribers.forEach(Subscriber::awaitTerminated);
+    }
+  }
+
+  @Override
+  protected void handlePermanentError(StatusException error) {
+    stop();
+  }
+
+  private void handleAssignment(Set<Partition> assignment) {
+    try {
+      try (CloseableMonitor.Hold h = monitor.enter()) {
+        if (shutdown) return;
+        Set<Partition> livePartitions = ImmutableSet.copyOf(liveSubscriberMap.keySet());
+        for (Partition partition : livePartitions) {
+          if (!assignment.contains(partition)) {
+            stopSubscriber(liveSubscriberMap.remove(partition));
```

Comment:
I'm a bit worried about acking the assignment after calling stopAsync, and not waiting for the Subscriber to actually stop. I understand that we now tear down the stream immediately, but I don't think we should rely on that externally, since it's an implementation detail.

Have you considered acking assignments asynchronously? It would be a bit more complex, but I think it's the right thing to do.

Reply:
I don't think it is the right thing to do. This isn't relying on the fact that we stop synchronously for correctness: All that matters is after stopAsync is called, we won't retry on the stream, so an eventual state in which stopAsync() set a "shutdown" boolean but didn't wait for the stream to stop is still correct.

Acking asynchronously adds some complexity that I don't think is needed since the work is bounded. WDYT?

Reply:
Yeah, I think that's fair. "All that matters is after stopAsync is called, we won't retry on the stream"; that still sounds like implementation details, but I agree that the alternative is a lot more complex.